### PR TITLE
Fixes #4716 Add more statistics for events

### LINKED ISSUE #4716: Add more statistics for events
**I'm submitting a ...**  (check one with "x")
- [ ] bug report
- [x] feature request
- [ ] support request => Please do not submit support requests here, instead ask your query in out Gitter channel at https://gitter.im/fossasia/open-event-server

**Current behavior:**

Currently, admin statistics is limited to 
1. `Event` how much live, past or draft
2. `Users` for all users etc

**Expected behavior:**

Next step can be stats for
1. `Event` how much belongs to a specific event-type and event-topic or even event-subtopics.
2. `Event` how much events are paid or free.
3. `Event` how much events are there in a specific state, city or country.

**Steps to reproduce:**

Not required. As its a feature.

**Related code:**

```
insert any relevant code here else remove this section
```

**Other information:**

**System information:** 

```
Your operating system
```

```
output of `python --version`
```
**Wanna work on this issue**

@niranjan94 thoughts please ?

### PR DESCRIPTION
**Pull request by bhaveshAn**
Fixes #4716 Add more statistics for events

#### Checklist

- [x] I have read the [Contribution & Best practices Guide](https://blog.fossasia.org/open-source-developer-guide-and-best-practices-at-fossasia) and my PR follows them.
- [x] My branch is up-to-date with the Upstream `development` branch.
- [ ] The unit tests pass locally with my changes 
- [ ] I have added tests that prove my fix is effective or that my feature works
- [x] I have added necessary documentation (if appropriate)

- [ ] All the functions created/modified in this PR contain relevant docstrings.

#### Short description of what this resolves:

Adds Event Type and Event Topic Statistics Details

#### Changes proposed in this pull request:

- As of now, added Event Type and Topic Statistics Detail at following APIs
`/v1/admin/statistics/event-types`
`/v1/admin/statistics/event-topics`
-
-

@niranjan94 @iamareebjamal @shubham-padia Please review thanks.

### CODE REVIEW

**Comment by bhaveshAn** [comment]
@niranjan94 @CosmicCoder96  Please help to fix the travis build.

**Comment by poush** [comment]
You travis build error is due to mismatch of the json response from api with the response present in api blueprint.

File: app/api/admin_statistics_api/events.py [hunk 1]
```diff
@@ -50,3 +53,405 @@ class AdminStatisticsEventDetail(ResourceDetail):
...
+    @staticmethod
+    def camp_treat_retreat_count(obj):
+        try:
+            events_type = EventType.query.filter_by(name='Camp, Treat & Retreat').one()
+        except NoResultFound:
+            return 0
+        return get_count(Event.query.filter_by(event_type_id=events_type.id))
+
+    @staticmethod
+    def appearance_or_signing_count(obj):
```
[the remark this answers is not in the document]
Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: Its because marshmallow provides Method field `fields.Method()` which does not contain any arguments other then obj 
Please see under Method Fields section at http://marshmallow.readthedocs.io/en/latest/custom_fields.html

File: app/api/admin_statistics_api/events.py [hunk 2]
```diff
@@ -50,3 +53,405 @@ class AdminStatisticsEventDetail(ResourceDetail):
...
+
+    rally = fields.Method("rally_count")
+    screening = fields.Method("screening_count")
+    seminar_or_talk = fields.Method("seminar_or_talk_count")
+    tour = fields.Method("tour_count")
+    tournament = fields.Method("tournament_count")
+    other = fields.Method("other_count")
+
+    @staticmethod
+    def camp_treat_retreat_count(obj):
```
Comment (srv-twry) [inline app/api/admin_statistics_api/events.py]: Can you explain why this method needs to be static ? If we look at the other Schema classes like ```AdminStatisticsEventSchema```, they aren't static.

Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: @srv-twry I have made it so because earlier codacy reported these error as to 
`Method could be a function`

@niranjan94 So should I retain it as `@staticmethod` or define the methods without `@staticmethod`

Comment (poush) [reply app/api/admin_statistics_api/events.py]: Static method don't receive `self` while the marshmallow's definition provides it. Therefore if the interface of marshmallow doesn't declare it as static then don't define it as static.

File: app/api/admin_statistics_api/events.py [hunk 3]
```diff
@@ -50,3 +53,405 @@ class AdminStatisticsEventDetail(ResourceDetail):
         'class': NoModelLayer,
         'session': db.session
     }
+
+
+class AdminStatisticsEventTypeSchema(Schema):
```
Comment (poush) [inline app/api/admin_statistics_api/events.py]: Like other APIs, this should be in the schema folder to make it readable and follow the standards.

**Comment by bhaveshAn** [comment]
@poush 
https://test16572.docs.apiary.io/#
Path 
1. Admin Statistics-> Event Types Statistics Detail
2. Admin Statistics-> Event Topics Statistics Detail

File: app/api/schema/admin_statistics_schema/event_topics.py [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        except NoResultFound:
+            return 0
+        return get_count(Event.query.filter_by(event_type_id=events_topic.id))
+
+    def other_count(self, obj):
+        try:
+            events_topic = EventTopic.query.filter_by(name='Other').one()
+        except NoResultFound:
+            return 0
+        return get_count(Event.query.filter_by(event_topic_id=events_topic.id))
```
[the remark this answers is not in the document]
Comment (srv-twry) [reply app/api/schema/admin_statistics_schema/event_topics.py]: Please fix these.

[the remark this answers is not in the document]
Comment (bhaveshAn) [reply app/api/schema/admin_statistics_schema/event_topics.py]: has been fixed otherwise it would have came with red mark at EOF

[the remark this answers is not in the document]
Comment (iamareebjamal) [reply app/api/schema/admin_statistics_schema/event_topics.py]: @bhaveshAn They are not fixed or else they would have shown as outdated

[the remark this answers is not in the document]
Comment (bhaveshAn) [reply app/api/schema/admin_statistics_schema/event_topics.py]: @iamareebjamal Here is the codacy report for the last commit https://app.codacy.com/app/mb/open-event-orga-server/pullRequest?prid=1643346
where it has reported only errors like `Method could be a function`
Please check.

**Comment by bhaveshAn** [comment]
@niranjan94 @SaptakS @poush Please review.
Is any test required.
Demo of api blueprint update 
https://test16572.docs.apiary.io/#
Path

    Admin Statistics-> Event Types Statistics Detail
    Admin Statistics-> Event Topics Statistics Detail

**Comment by bhaveshAn** [comment]
@niranjan94 I followed the already implemented `/v1/admin/statistics/events` API https://github.com/fossasia/open-event-server/blob/development/app/api/admin_statistics_api/events.py to add this implementation which would be a way to align other APIs in same way.
Also, to change the implementation, I think to use a marshmallow schema with **Json** 1.0 spec we should follow marshmallow way, otherwise using aggregate expressions provided by PostgreSQL can divert the implementation. 
Otherwise, it would lead to having SQL commands and use of ORM both on the application.
Thoughts please...

**Comment by schedutron** [comment]
@bhaveshAn Just curious, if PostgreSQL aggregate commands are not preferred by you, can redundancy be reduced by using marshmallow `context`, as I asked previously?

File: app/api/admin_statistics_api/events.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: [Example](http://marshmallow.readthedocs.io/en/latest/custom_fields.html) provided is returning whether a user is author the blog or not .
In our case, 
```
define function as 
def get_event_type_count(event_type_name):
    try:
        events_type = EventType.query.filter_by(name=event_type_name).one()
    except NoResultFound:
        return 0
    return get_count(Event.query.filter_by(event_type_id=events_type.id))

c_or_p =  get_event_type_count('Concert or Performance')
attr =  get_event_type_count('Attraction')

schema.context = {'Concert or Performance': c_or_p, 'Attraction': attr, and so on}

**In Schemas attributes**

concert_or_performance = fields.Function(lambda value, context: value == context['Concert or Performance':])
 
```

Not sure if it's write but expected @schedutron

**Comment by bhaveshAn** [comment]
@schedutron https://github.com/fossasia/open-event-server/pull/4724#discussion_r189680092

File: app/api/admin_statistics_api/events.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (poush) [reply app/api/admin_statistics_api/events.py]: I think it's very simple now, create a helper function in helpers and import it here. Now call it from lamda function like:
`concert_or_performance = fields.Function(lambda value: stats_helper_func('Concert or Performance'))`
 
I think this will do

**Comment by niranjan94** [comment]
@bhaveshAn in the existing case that you are talking about, there are only three states for events. Draft, published, past. But in this case, we also would be allowing users to add more categories/sub-categories. So, we need to use some sort of aggregate functions to efficiently get the count of each category. 

> it would lead to having SQL commands and use of ORM both on the application

If so, that is totally okay. An ORM cannot accomplish everything. It is completely okay to mix raw queries for such special cases alongwith an ORM in an application.

File: app/api/admin_statistics_api/events.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: Please review @poush

File: app/api/schema/admin_statistics_schema/event_types.py [hunk 5]
```diff
@@ -20,176 +17,24 @@ class Meta:
...
+    convention = fields.Function(lambda value: get_event_type_count(event_type_name='Convention'))
+    dinner_or_gala = fields.Function(lambda value: get_event_type_count(event_type_name='Dinner or Gala'))
+
+    festival_or_fair = fields.Function(lambda value: get_event_type_count(event_type_name='Festival or Fair'))
+    game_or_competition = fields.Function(lambda value: get_event_type_count(event_type_name='Game or Competition'))
+    meeting_or_networking_event = fields.Function(lambda value: get_event_type_count(event_type_name='Meeting or Networking Event'))
+    party_or_social_gathering = fields.Function(lambda value: get_event_type_count(event_type_name='Party or Social Gathering'))
+    race_or_endurance_event = fields.Function(lambda value: get_event_type_count(event_type_name='Race or Endurance Event'))
+
+    rally = fields.Function(lambda value: get_event_type_count(event_type_name='Rally'))
```
Comment (schedutron) [inline app/api/schema/admin_statistics_schema/event_types.py]: Why is `value` being taken as input to the `lambda` functions?

**Comment by poush** [comment]
@niranjan94 @SaptakS @iamareebjamal Can you review the current implementation of creating helper function and settings to marshmallow schemas?

**Comment by niranjan94** [comment]
@bhaveshAn @poush my original concern still remains

> But in this case, we also would be allowing users to add more categories/sub-categories. (See https://github.com/fossasia/open-event-frontend/pull/1159)

We will have to allow the users to add/remove/edit topics/sub-topics. So, the API response needs to be dynamic.

**Comment by bhaveshAn** [comment]
@Kreijstal @niranjan94 @poush @SaptakS  @maxlorenz Please review.

> We will have to allow the users to add/remove/edit topics/sub-topics. So, the API response needs to be dynamic.

Did it @niranjan94 Please review.

File: app/api/admin_statistics_api/events.py [hunk 6]
```diff
@@ -1,42 +1,37 @@
...
+        event_topic_counts[event_topic.name] = value
+    return jsonify(event_topic_counts)
+
+
+@event_statistics.route('/event-types', methods=['GET'])
+@jwt_required()
+def event_types_count():
+    event_types_counts = {}
+    for event_type in EventType.query.all():
+        value = get_count(Event.query.filter_by(event_topic_id=event_type.id))
```
Comment (niranjan94) [inline app/api/admin_statistics_api/events.py]: @bhaveshAn like I have been telling from the very beginning .. Why couldn't you use Postgres's query to group and aggregate ? 

Here, if there are 20 event types, you will be making 21 queries to get this statistics when all you need is a single query. I'm not sure why this is hard to understand.

Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: @niranjan94 Using query:
```
SELECT event_topics.name, count(events.event_topic_id) FROM events, event_topics WHERE events.event_topic_id=event_topics.id GROUP BY event_topics.name;
```
I am only getting the event topic which has a event and not 0 when there isn't any event of a event topic.
Is something wrong with the query?

> Here, if there are 20 event types, you will be making 21 queries to get this statistics when all you need is a single query. I'm not sure why this is hard to understand.

Earlier I mean the focus is on to get the dynamic response. That's why it happened.

Comment (niranjan94) [reply app/api/admin_statistics_api/events.py]: Hi @bhaveshAn 

You should be using a `JOIN` query. Try this one. Should work.

```sql
SELECT event_types.id as event_type_id, event_types.name as event_type, COUNT(events.id) AS count
FROM event_types
LEFT JOIN events
ON events.event_type_id = event_types.id
GROUP BY event_types.id
```

File: app/api/admin_statistics_api/events.py [hunk 7]
```diff
@@ -1,42 +1,42 @@
...
+@jwt_required()
+def event_topic_count():
+    result_set = db.engine.execute(text(
+        "SELECT event_topics.name, COUNT(events.id) FROM event_topics " +
+        "LEFT JOIN events ON events.event_topic_id = event_topics.id " +
+        "GROUP BY event_topics.id;"
+    ))
+    event_topics_counts = {}
+    for each in list(result_set):
+        event_topics_counts[each[0]] = each[1]
```
Comment (iamareebjamal) [inline app/api/admin_statistics_api/events.py]: event_topics_counts = dict(list(result_set))

File: app/api/admin_statistics_api/events.py [hunk 8]
```diff
@@ -1,42 +1,42 @@
...
+    return (jsonify(event_topics_counts))
```
Comment (iamareebjamal) [inline app/api/admin_statistics_api/events.py]: Extra parens

File: app/api/admin_statistics_api/events.py [hunk 9]
```diff
@@ -20,9 +20,8 @@ def event_topic_count():
         "GROUP BY event_topics.id;"
     ))
     event_topics_counts = {}
-    for each in list(result_set):
-        event_topics_counts[each[0]] = each[1]
-    return (jsonify(event_topics_counts))
+    event_topics_counts = dict(list(result_set))
```
Comment (iamareebjamal) [inline app/api/admin_statistics_api/events.py]: Reassigning is not needed

File: app/api/admin_statistics_api/events.py [hunk 10]
```diff
@@ -1,42 +1,40 @@
...
+@event_statistics.route('/event-topics', methods=['GET'])
+@jwt_required()
+def event_topic_count():
+    result_set = db.engine.execute(text(
+        "SELECT event_topics.name, COUNT(events.id) FROM event_topics " +
+        "LEFT JOIN events ON events.event_topic_id = event_topics.id " +
+        "GROUP BY event_topics.id;"
+    ))
+    event_topics_counts = {}
+    event_topics_counts = dict(list(result_set))
```
Comment (iamareebjamal) [inline app/api/admin_statistics_api/events.py]: No need of reassigning

Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: Mean removing `event_topics_counts = {}` ?

File: app/api/admin_statistics_api/events.py (hunk 6 above)
Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: Hi @niranjan94 Please review now.

**Comment by bhaveshAn** [comment]
Hi @niranjan94 @poush @iamareebjamal @srv-twry @maxlorenz @schedutron Please review.
Dynamic response of API using SQL commands.
Added Dredd tests as well.

File: docs/api/api_blueprint.apib [hunk 11]
```diff
@@ -20828,6 +20828,90 @@ Get the task status and the final result of the task as response
         }
 
 
+## Event Types Statistics Details [/v1/admin/statistics/event-types]
+
+**Event Types:**
+
+| Parameter | Description | Type |
```
Comment (niranjan94) [inline docs/api/api_blueprint.apib]: Do these parameters types match the response ? Can you post a sample response here ?

Comment (bhaveshAn) [reply docs/api/api_blueprint.apib]: @niranjan94 
For **EventTypes**`/v1/admin/statistics/event-types`

screenshot from 2018-06-09 15-15-49

For **EventTopics**`/v1/admin/statistics/event-topics` 
	
screenshot from 2018-06-09 15-15-35

Comment (niranjan94) [reply docs/api/api_blueprint.apib]: @bhaveshAn please make the response something like,

```js
[
    {
        name: "Appearance or Signing",
        id: 1,
        count: 12
    },
    {
        name: "Attraction",
        id: 2,
        count: 1
    }
]
```

That way the clients have access to the ID of the type too. This would allow clients to implement features such as viewing all events of a specific type right from the statistics page

Comment (bhaveshAn) [reply docs/api/api_blueprint.apib]: @niranjan94 Done, Please review.

**EventTypes**
screenshot from 2018-06-09 18-25-23
**EventTopics**
screenshot from 2018-06-09 18-25-15

File: app/api/admin_statistics_api/events.py [hunk 12]
```diff
@@ -15,24 +15,38 @@
...
-    event_topics_counts = dict(list(result_set))
-    return jsonify(event_topics_counts)
+    event_topics_count = []
+    for each in list(result_set):
+        event_topic = {}
+        event_topic["name"] = each[0]
+        event_topic["id"] = each[1]
+        event_topic["count"] = each[2]
+        event_topics_count.append(event_topic)
+    return jsonify(event_topics_count)
```
Comment (iamareebjamal) [inline app/api/admin_statistics_api/events.py]: Use the approach suggested previously

Comment (iamareebjamal) [reply app/api/admin_statistics_api/events.py]: Sure

Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: Done @iamareebjamal Please review.

screenshot from 2018-06-09 18-25-23
screenshot from 2018-06-09 18-25-15

File: app/api/admin_statistics_api/events.py [hunk 13]
```diff
@@ -39,14 +34,9 @@ def event_types_count():
...
     ))
-    event_types_count = []
-    for each in list(result_set):
-        event_type = {}
-        event_type["name"] = each[0]
-        event_type["id"] = each[1]
-        event_type["count"] = each[2]
-        event_types_count.append(event_type)
-    return jsonify(event_types_count)
+    event_types_counts = []
```
Comment (iamareebjamal) [inline app/api/admin_statistics_api/events.py]: No need to reassign

File: app/api/admin_statistics_api/events.py [hunk 14]
```diff
@@ -20,14 +20,9 @@ def event_topic_count():
...
     ))
-    event_topics_count = []
-    for each in list(result_set):
-        event_topic = {}
-        event_topic["name"] = each[0]
-        event_topic["id"] = each[1]
-        event_topic["count"] = each[2]
-        event_topics_count.append(event_topic)
-    return jsonify(event_topics_count)
+    event_topics_counts = []
```
Comment (iamareebjamal) [inline app/api/admin_statistics_api/events.py]: No need to reassign

File: app/api/admin_statistics_api/events.py (hunk 13 above)
Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: resolved

File: app/api/admin_statistics_api/events.py (hunk 14 above)
Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: resolved. please review.

**Comment by bhaveshAn** [comment]
@iamareebjamal @niranjan94 Can this PR be reviewed and merged if the goal is completed.
screenshot from 2018-06-09 18-25-23
screenshot from 2018-06-09 18-25-15

File: app/api/admin_statistics_api/events.py [hunk 15]
```diff
@@ -1,42 +1,40 @@
...
-        return get_count(Event.query.filter_by(state='published'))
-
-    def events_past_count(self, obj):
-        return get_count(Event.query.filter(Event.ends_at < datetime.now(pytz.utc)))
+from app.api.schema.admin_statistics_schema.events import AdminStatisticsEventSchema
+
+event_statistics = Blueprint('event_statistics', __name__, url_prefix='/v1/admin/statistics')
+
+
+@event_statistics.route('/event-topics', methods=['GET'])
```
Comment (Kreijstal) [inline app/api/admin_statistics_api/events.py]: Why is this done here instead on the __init__ file of the api?

File: docs/api/api_blueprint.apib [hunk 16]
```diff
@@ -20828,6 +20828,88 @@ Get the task status and the final result of the task as response
...
+
+    + Headers
+
+            Accept: application/vnd.api+json
+
+            Authorization: JWT <Auth Key>
+
++ Response 200 (application/json)
+
+        {}
```
Comment (Kreijstal) [inline docs/api/api_blueprint.apib]: Describe the type of the response!

**Comment by Kreijstal** [comment]
But most importantly, why are you creating a complete new route /admin/statistics instead of implementing it through flask-rest-jsonapi? It could be done with a ResouceDetail couldn't it?

**Comment by bhaveshAn** [comment]
@Kreijstal 

> Why are you creating a complete new route /admin/statistics instead of implementing it through flask-rest-jsonapi? It could be done with a ResouceDetail couldn't it?

No, ResouceDetail is used to update get and delete a record in database and requires ID to do so. Here, we are doing statistics of our data from event_type, event_topic and event models.
These are some special cases when we have to use Flask blueprint to add API route. You can take reference from the APIs of 
1. `upload_routes`
2. `export_routes`
3. `import_routes`
4. `celery_routes`
5. `auth_routes`
6. `event_copy`
from https://github.com/fossasia/open-event-server/blob/development/app/__init__.py#L110

File: app/api/admin_statistics_api/events.py (hunk 15 above)
Comment (bhaveshAn) [reply app/api/admin_statistics_api/events.py]: init is used to add route for REST APIs made by Flast Rest JSONAPI.

**Comment by bhaveshAn** [comment]
@iamareebjamal @niranjan94 Can this PR be reviewed and merged if the goal is completed.
screenshot from 2018-06-09 18-25-23
screenshot from 2018-06-09 18-25-15

**Comment by iamareebjamal** [comment]
@srv-twry @schedutron @maxlorenz Please review

File: docs/api/api_blueprint.apib [hunk 17]
```diff
@@ -20834,6 +20834,88 @@ Get the task status and the final result of the task as response
...
```
Comment (srv-twry) [inline docs/api/api_blueprint.apib]: How about adding a Factory for testing and then adding the expected response here ? Empty response doesn't server the purpose IMO.

File: app/api/schema/admin_statistics_schema/events.py [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+    past = fields.Method("events_past_count")
+
+    def events_draft_count(self, obj):
+        return get_count(Event.query.filter_by(state='draft'))
+
+    def events_published_count(self, obj):
+        return get_count(Event.query.filter_by(state='published'))
+
+    def events_past_count(self, obj):
+        return get_count(Event.query.filter(Event.ends_at < datetime.now(pytz.utc)))
```
[the remark this answers is not in the document]
Comment (srv-twry) [reply app/api/schema/admin_statistics_schema/events.py]: I see that 'datetime' and 'pytz' are fixed but there is no newline at the end of the file. Please add one empty line at the end.

File: docs/api/api_blueprint.apib [hunk 19]
```diff
@@ -20834,6 +20834,88 @@ Get the task status and the final result of the task as response
...
+| `Screening`  | No. of events of type - Screening | Integer |
+| `Game or Competition`  | No. of events of type - Game or Competition | Integer |
+| `Party or Social Gathering`  | No. of events of type - Party or Social Gathering | Integer |
+| `Race or Endurance Event`  | No. of events of type - Race or Endurance Event | Integer |
+| `Meeting or Networking Event`  | No. of events of type - Meeting or Networking Event | Integer |
+| `Attraction`  | No. of events of type - Attraction | Integer |
+| `Class, Training, or Workshop`  | No. of events of type - Class, Training, or Workshop | Integer |
+| `Appearance or Signing`  | No. of events of type - Appearance or Signing | Integer |
+| `Tournament`  | No. of events of type - Tournament | Integer |
+| `Rally`  | No. of events of type - Rally | Integer |
```
Comment (niranjan94) [inline docs/api/api_blueprint.apib]: This schema doesn't match with what is there right now. Also, you don't have to mention the actual topics here in the doc. You can just put an example response maybe. Since these topics/sub-topics etc are dynamic.

**Comment by bhaveshAn** [comment]
@niranjan94 @iamareebjamal Please review and merge. If possible. Thanks!

File: docs/api/api_blueprint.apib (hunk 19 above)
Comment (bhaveshAn) [reply docs/api/api_blueprint.apib]: Done

File: docs/api/api_blueprint.apib (hunk 17 above)
Comment (bhaveshAn) [reply docs/api/api_blueprint.apib]: Since response is dynamic. we cannot add factory here. As which event type/topic and how many of them will be returned.

> def events_past_count(self, obj):
        return get_count(Event.query.filter(Event.ends_at < datetime.now(pytz.utc)))

> Please add one empty line at the end.

There is empty line. If that wasn't there then it is always shown with a red mark

**Comment by bhaveshAn** [comment]
Admin can Rebase or squash while merging.

**Comment by niranjan94** [comment]
@bhaveshAn 

> Admin can Rebase or squash while merging.

Yes. Admin can. Be I believe it is the contributors responsibility to make sure that their PR is as perfect as possible. So, please squash the commits.

**Comment by bhaveshAn** [comment]
@niranjan94 
Yes, I mean i can't as i have merged commits from upstream/development here from github ui.
Please merge this one. 
My bad, I will take care from next time.

**Comment by iamareebjamal** [comment]
@bhaveshAn Not everything can be done using GitHub GUI. Please use git CLI

**Comment by srv-twry** [comment]
@bhaveshAn 
Checkout the branch locally:
```
git fetch upstream
git rebase upstream/development
git reset --soft HEAD~x && git commit -m "commit message"
git push -f origin branch_name
```

x: Number of commits in the PR.(Use `git log` and count them manually)
I have assumed you've named the main remote repo as `upstream` according to conventions.
